[WIP] CSV datatype

vavrusa opened the pull request:
This is another attempt at the CSV type support, I'd like to know your thoughts before I move forward with it.
## Fixes
- made a superclass out for Tabular (TabularData) with metadata and generic methods
- made a CSV class for CSV-style tabular data
- `stats/gsummary` works with [RPy2](http://rpy.sourceforge.net/rpy2/doc-dev/html/overview.html) classic mode, as the original RPy is [pretty much dead](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=705847)
## How to test

I have attached some example files at https://gist.github.com/vavrusa/f8bb8d059c6c96dc7585
to test it, the one with the first line as a header should be sniffed correctly when uploading, the rest should be correctly displayed when saved as `tabular`. Visualization for all should be correct.

vavrusa committed:
stats/gsummary: RPy is deprecated, use RPy2/classic as an alternative

vavrusa committed:
datatypes: extracted TabularData superclass from Tabular, CSV subclass

jmchilton commented on lib/galaxy/datatypes/tabular.py:
```diff
@@ -20,17 +21,209 @@
+    def column_dataprovider( self, dataset, **settings ):
+        """Uses column settings that are passed in"""
+        dataset_source = dataproviders.dataset.DatasetDataProvider( dataset )
+        delimiter = dataset.metadata.delimiter
```
Overall this seems pretty solid to me - can you create a tabular file on master (which won't have the `delimiter` metadata defined), then checkout this branch and verify these column providers still work with the old datasets. I think there are other places we are defensive for older datasets and newer metadata - I just wanted to make sure you have confirmed that isn't needed here.

jmchilton replied:
Chatted with @blankenberg about this - it seems that this is what `no_value=` does in the MetadataElement definition about. So I guess the `no_value` for `delimiter` should be `\t`.

vavrusa replied:
Hey, I just checked - to me it looks okay (file: [csv-tab-commentheader-csv](https://gist.github.com/vavrusa/f8bb8d059c6c96dc7585#file-csv-tab-commentheader-csv))

vavrusa commented on config/plugins/visualizations/charts/config/charts.xml:
```diff
@@ -5,6 +5,7 @@
         <data_source>
             <model_class>HistoryDatasetAssociation</model_class>
             <test type="isinstance" test_attr="datatype" result_type="datatype">tabular.Tabular</test>
+            <test type="isinstance" test_attr="datatype" result_type="datatype">tabular.CSV</test>
```
I have a question here - why is it not possible to specify superclass that isn't in the the file extension registry?
Is it designed like this for a reason or just a shortcoming?

carlfeberhard replied:
The visualization framework calls the registry to lookup the datatype class by name/string.

Generally the superclasses have had less in common with each other as far as information contained than their format or methods for sniffing AFAIK (e.g. Binary, Text, etc.) with some exceptions (Interval). So, in effect, it hasn't been needed (again, AFAIK).

jmchilton commented:
Yeah - I just check it and somehow it is getting '\t' as the delimiter even though the `no_value=[]`. Still think the `no_value="\t"` would be a better default though for that metadata element.

So does the scatterplot work for you? I don't get any visualizations enabled except IPython if I upload a `csv` file.

jmchilton commented:
Haven't heard any news on this for a couple weeks - any chance for status update? I am going to merge #204 soon unless there is a strong objection which will introduce more conflicts to this.

vavrusa commented:
Yeah, the notification slipped my mind. The scatterplots works for me, maybe it's not in your configuration? The thing is, that everything in `config/` is blocked by https://github.com/galaxyproject/galaxy/blob/dev/.gitignore#L65 (which is a bug btw.), even the `config/plugins/visualizations/charts/config/charts.xml`, so maybe it didn't update for you?

dannon committed:
Merge branch 'guerler-workflow_form_000' into dev

jmchilton committed:
Merge remote-tracking branch 'vavrusa/table-datatype' into dev.

Conflicts:
	lib/galaxy/datatypes/tabular.py

jmchilton committed:
Fix sample datatype definition for CSV.

For me it seems like the viz plugins only detect the datatype if the plugin datatype (e.g. ``tabular.CSV``) is explicitly instantiated by some extension directly. In your sample config you have ``subclass="True"`` so ``tabular.CSV`` wasn't working. This is probably a bug with the datatype code - but nonetheless I think dropping subclass from the sample is an okay workaround - because it does seem to be a direct mapping.

jmchilton commented:
Figured out the viz problem - I think a small tweak to the sample config fixes it. I have opened a PR against your branch that resolves this and merges in the latest dev so the PR reapplies cleanly.

vavrusa committed:
Merge pull request #1 from jmchilton/table

Updates for PR #191 (checked the 508194f)

martenson commented:
Thank you for the contribution @vavrusa and @jmchilton for the review.

vavrusa commented:
:tada: thanks!
